### DISCUSSION TITLE
Support for S3 Access via EKS Service Accounts

### DISCUSSION THREAD

**Discussion by be-aws-architect** [directus/directus#24689] [Feature Requests]
### Describe the Improvement

Currently it's only possible to setup an integration with S3 using Access Keys, an outdated setup which most applications have replaced with using IAM roles linked to EKS Service Accounts. Allow the integration to assume an IAM Role for S3 access instead of using hardcoded Access Keys that need to be manually configured and rotated.
[https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html](https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html)

**Comment by rijkvanzanten** [comment]
Quick note: Restack is an AI generated slob that's not official, fact checked, or even accurate most of the time. They're stealing our content, rewrite it with AI, and then launch it with aggressive SEO to get views 😐

**Comment by rafalkrupinski** [comment]
> most applications have replaced with using IAM roles linked to EKS Service Accounts

source for the claim?

**Comment by be-aws-architect** [comment]
https://signoz.io/docs/tutorial/s3-integration-iam-role-eks/

https://milvus.io/docs/s3.md

These are just some public examples I found with some quick googling, however most of my experience comes from closed-source software of customers I serve at my MSP which I can't share of course.

**Comment by rafalkrupinski** [comment]
AWS is not the only S3 implementation these days, are you sure they all support roles?

**Comment by be-aws-architect** [comment]
Hi,

What do you mean by that? Other file storage implementations?

Mvg,

Tommy Veevaete
________________________________

**Comment by rafalkrupinski** [comment]
For example Ceph, a S3 implementation.

Also, you can be connecting to S3 from outside of AWS, so I'd rather directus not _switch_ to IAM roles for S3.

**Comment by be-aws-architect** [comment]
Most certainly 3rd party custom storage implementations are even more difficult to support...

There already is support for S3, the next logical step is to secure the authentication to it.

Mvg,

Tommy Veevaete
________________________________

**Comment by rijkvanzanten** [comment]
@rafalkrupinski fwiw we wouldn't drop support for access id / secret key for that exact reason 🙂 I use DigitalOcean Spaces a lot for my own stuff which is S3-compatible, but there's no IAM roles as you'd find on AWS there

**Comment by rijkvanzanten** [comment]
I won't be able to find the time to dive into this one myself as it's a bit more niche, but I'll more than happily take a PR. We use AWS' official S3-sdk for JS, so hopefully supporting EKS should be possible ootb.

We define all the configuration parameters here:

https://github.com/directus/directus/blob/28aaf739ba75980f4cb5ed1fa8c31b900dd97765/packages/storage-driver-s3/src/index.ts#L41-L58

and then use them to create the aws s3 sdk client here:

https://github.com/directus/directus/blob/28aaf739ba75980f4cb5ed1fa8c31b900dd97765/packages/storage-driver-s3/src/index.ts#L81-L132

@be-aws-architect wanna take a stab at a PR? 🙂
